frame/alliance/src/mock.rs

```diff
@@ -184,24 +194,33 @@ impl ProposalProvider<AccountId, H256, RuntimeCall> for AllianceProposalProvider
 
 	fn vote_proposal(
 		who: AccountId,
-		proposal: H256,
+		proposal_bounded: Bounded<RuntimeCall>,
 		index: ProposalIndex,
 		approve: bool,
 	) -> Result<bool, DispatchError> {
-		AllianceMotion::do_vote(who, proposal, index, approve)
+		AllianceMotion::do_vote(who, proposal_bounded, index, approve)
 	}
 
 	fn close_proposal(
-		proposal_hash: H256,
+		proposal_bounded: Bounded<RuntimeCall>,
 		proposal_index: ProposalIndex,
 		proposal_weight_bound: Weight,
 		length_bound: u32,
 	) -> DispatchResultWithPostInfo {
-		AllianceMotion::do_close(proposal_hash, proposal_index, proposal_weight_bound, length_bound)
+		AllianceMotion::do_close(
+			proposal_bounded,
+			proposal_index,
+			proposal_weight_bound,
+			length_bound,
+		)
+	}
+
+	fn proposal_of(proposal_bounded: Bounded<RuntimeCall>) -> Option<RuntimeCall> {
+		AllianceMotion::proposal_of(proposal_bounded)
 	}
 
-	fn proposal_of(proposal_hash: H256) -> Option<RuntimeCall> {
-		AllianceMotion::proposal_of(proposal_hash)
+	fn bound_proposal(proposal: RuntimeCall) -> Result<Bounded<RuntimeCall>, DispatchError> {
+		Preimage::bound(proposal)
```

Comment:
This emits `Event::Noted { hash }` and writes to storage if not already noted. 
I think there should be a method to get the bounded version without writing or emitting.

Reply:
How can you bound it without writing it to storage?  
And if it writes to storage then it should emit an event to notify the outside world.

Reply:
> How can you bound it without writing it to storage?

If the length exceeds, then compute the hash, without storing. The `bound_no_store` method I was talking about in [this comment](https://github.com/paritytech/substrate/pull/14355#discussion_r1251714214).
It depends on how the `bound_proposal` method is used, and as far as I remember, it is only to retrieve the value from storage.